src/operator/tensor/indexing_op.h

```diff
@@ -493,6 +493,133 @@ void BatchTakeOpForward(const nnvm::NodeAttrs& attrs,
   });
 }
 
+/*!
+ * \brief The parameters of the one_hot operator.
+ */
+struct OneHotParam : public dmlc::Parameter<OneHotParam> {
+  int depth;
+  double on_value;
+  double off_value;
+  int axis;
+  int dtype;
+  DMLC_DECLARE_PARAMETER(OneHotParam) {
+    DMLC_DECLARE_FIELD(depth)
+      .describe("The dimension size at dim = axis.");
+    DMLC_DECLARE_FIELD(on_value)
+      .set_default(1.0f)
+      .describe("The value assigned to the locations represented by indices.");
+    DMLC_DECLARE_FIELD(off_value)
+      .set_default(0.0f)
+      .describe("The value assigned to the locations not represented by indices.");
+    DMLC_DECLARE_FIELD(axis)
+      .set_default(-1)
+      .describe("The dimension position at which a new dimension will be created."
+                " So far, only axis = -1 is supported, which means a new dimension"
+                " is always appended at the end of the array. Setting it to other"
+                " numbers will cause the program to throw an exception");
+    DMLC_DECLARE_FIELD(dtype)
+      .set_default(mshadow::kFloat32)
+      .add_enum("float32", mshadow::kFloat32)
+      .add_enum("float64", mshadow::kFloat64)
+      .add_enum("float16", mshadow::kFloat16)
+      .add_enum("uint8", mshadow::kUint8)
+      .add_enum("int32", mshadow::kInt32)
+      .describe("DType of the output");
+  }
+};
+
+inline void GetOneHotParams(const OneHotParam& param, int* depth, double* on_value,
+                            double* off_value, int* axis, int* dtype) {
+  *depth = param.depth;
+  CHECK_GE(*depth, 0) << "Dimension size, depth, must be a non-negative integer";
+  *on_value = param.on_value;
+  *off_value = param.off_value;
+  *axis = param.axis;
+  CHECK_EQ(*axis, -1) << "Only support axis = -1 for now";
+  *dtype = param.dtype;
+}
+
+inline bool OneHotOpShape(const nnvm::NodeAttrs& attrs,
+                          std::vector<TShape> *in_attrs,
+                          std::vector<TShape> *out_attrs) {
+  const OneHotParam& param = nnvm::get<OneHotParam>(attrs.parsed);
+  CHECK_EQ(in_attrs->size(), 1);
+  CHECK_EQ(out_attrs->size(), 1);
+  // The shape of indices
+  const TShape& ishape = (*in_attrs)[0];
+
+  int depth = 0;
+  double on_value = 1.0;
+  double off_value = 0.0;
+  int axis = -1;
+  int dtype = mshadow::kFloat32;
+  GetOneHotParams(param, &depth, &on_value, &off_value, &axis, &dtype);
```

Comment:
GetOneHotParams provides a centralized place to do parameter check to ensure they are valid. Since params are used in shape, type, forward, and backward functions, this function avoids duplicating error checking code.